streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1498,6 +1516,37 @@ public synchronized boolean close(final Duration timeout) throws IllegalArgument
         return close(timeoutMs);
     }
 
+    /**
+     * Shutdown this {@code KafkaStreams} by signaling all the threads to stop, and then wait up to the timeout for the
+     * threads to join.
+     * @param options  contains timeout to specify how long to wait for the threads to shutdown, and a flag leaveGroup to
+     *                 trigger consumer leave call
+     * @return {@code true} if all threads were successfully stopped&mdash;{@code false} if the timeout was reached
+     * before all threads stopped
+     * Note that this method must not be called in the {@link StateListener#onChange(KafkaStreams.State, KafkaStreams.State)} callback of {@link StateListener}.
+     * @throws IllegalArgumentException if {@code timeout} can't be represented as {@code long milliseconds}
+     */
+    public synchronized boolean close(final CloseOptions options) throws IllegalArgumentException {
+        final String msgPrefix = prepareMillisCheckFailMsgPrefix(options.timeout, "timeout");
+        final long timeoutMs = validateMillisecondDuration(options.timeout, msgPrefix);
+        if (timeoutMs < 0) {
+            throw new IllegalArgumentException("Timeout can't be negative.");
+        }
+
+        if (options.leaveGroup) {
+            log.debug("Sending leave group trigger to removing instance from consumer group");
+            //removing instance from consumer group
+            adminClient.removeMembersFromConsumerGroup(
```

Comment:
Should we also wait on the returned result of remove-member to complete within the `timeoutMs` as well?

Reply:
if we do so, then should we pass the same `timeoutMs` to the close function? or should we subtract the time taken by the future to get resolved?

Meanwhile, I didn’t add a wait for the future to get resolved, as I thought that if the call resolves then it would be removed from the consumer group and, if the call rejects or returns failure then after `session.timeout` the instance will be anyway removed from the group.

Reply:
Hi @sayantanu-dey 

1) Yeah I agree that if we wait for the future to get resolved before calling `consumer.close` it may be rejected. Actually, I think we should call `close` first, then using the admin to call remove-member.

2) What I meant is that we should consider only returning this `KafkaStreams#close` call after the future from the remove-member has completed, but also bounded by the timeoutMs. I.e. after the inner `close` returns we subtract the time taken and then use the remaining to wait for the future to get resolved. WDYT?

Reply:
Actually I got it wrong at the beginning :P The `RemoveMembersFromConsumerGroupOptions` only allows use to remove the static members, via the `groupInstanceId`, which is a static config not a dynamic value, so we just need to pass in that from the config, and not need to rely on `KafkaConsumer#groupMetadata`.

Sorry for causing the confusion..

Reply:
That makes sense, thanks.